File: backend/src/main/java/com/project2/domain/member/service/AuthService.java
package com.project2.domain.member.service;

import com.project2.domain.member.entity.Member;
import com.project2.domain.member.repository.MemberRepository;
import com.project2.global.exception.ServiceException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final MemberRepository memberRepository;
    private final AuthTokenService authTokenService;

    public Optional<Member> getMemberByAccessToken(String accessToken) {

        Map<String, Object> payload = authTokenService.getPayload(accessToken);

        if (payload == null) {
            return Optional.empty();
        }

        long id = (long) payload.get("id");
        String email = (String) payload.get("email");

        return Optional.of(
                Member.builder()
                        .id(id)
                        .email(email)
                        .build()
        );
    }

    public Optional<Member> getMemberByRefreshToken(String refreshToken) {
        Map<String, Object> payload = authTokenService.getPayload(refreshToken);

        if (payload == null) {
            return Optional.empty();
        }

        long id = (long) payload.get("id");

        return memberRepository.findById(id);
    }

    public Member getMemberByRefreshTokenOrThrow(String refreshToken) {
        return getMemberByRefreshToken(refreshToken)
                .orElseThrow(() -> new ServiceException("401", "유효하지 않은 리프레시 토큰이거나 회원을 찾을 수 없습니다."));
    }

    public String genAccessToken(Member member) {
        return authTokenService.genAccessToken(member);
    }

    public String genRefreshToken(Member member) {
        return authTokenService.genRefreshToken(member.getId());
    }

    public void validateOwner(long actorId, long memberId, String errorMessage) {
        if (actorId != memberId) {
            throw new ServiceException("403", errorMessage);
        }
    }
}

File: backend/src/main/kotlin/com/project3/domain/member/service/AuthService.kt
package com.project3.domain.member.service

import com.project3.domain.member.entity.Member
import com.project3.domain.member.repository.MemberRepository
import com.project3.global.exception.ServiceException
import org.springframework.stereotype.Service
import java.util.*

@Service
class AuthService(
        private val memberRepository: MemberRepository,
        private val authTokenService: AuthTokenService
) {

    fun getMemberByAccessToken(accessToken: String): Optional<Member> {
        val payload = authTokenService.getPayload(accessToken) ?: return Optional.empty()

        val id = payload["id"] as Long
        val email = payload["email"] as String

        return Optional.of(
                Member().apply {
                    this.id = id
                    this.email = email
                }
        )
    }

    fun getMemberByRefreshToken(refreshToken: String): Optional<Member> {
        val payload = authTokenService.getPayload(refreshToken) ?: return Optional.empty()
        val id = payload["id"] as Long

        return memberRepository.findById(id)
    }

    fun getMemberByRefreshTokenOrThrow(refreshToken: String): Member {
        return getMemberByRefreshToken(refreshToken)
                .orElseThrow { ServiceException("401", "유효하지 않은 리프레시 토큰이거나 회원을 찾을 수 없습니다.") }
    }

    fun genAccessToken(member: Member): String {
        return authTokenService.genAccessToken(member)
    }

    fun genRefreshToken(member: Member): String {
        return authTokenService.genRefreshToken(member.id!!)
    }

    fun validateOwner(actorId: Long, memberId: Long, errorMessage: String) {
        if (actorId != memberId) {
            throw ServiceException("403", errorMessage)
        }
    }
}
